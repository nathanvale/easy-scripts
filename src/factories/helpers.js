function createDoItForYouPrompt({type = 'select', name, message} = {}) {
  return {
    type,
    name,
    message,
    choices: [
      {
        title: `"No thank you. I can do that myself."`,
        value: false,
      },
      {
        title: `"Yes please! That would be awesome! :)"`,
        value: true,
      },
    ],
    initial: 1,
  }
}

module.exports = {
  createDoItForYouPrompt,
}
