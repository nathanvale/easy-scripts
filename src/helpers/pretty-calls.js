function prettyCalls(calls) {
  const argumentMapper = (item, argIndex) => `Argument ${argIndex + 1}:
    ${item}`
  const reducer = (previousValue, currentValue, currentCallIndex) => `${
    previousValue === '' ? previousValue : `${previousValue}\n`
  }Call ${currentCallIndex + 1}:
  ${currentValue.map(argumentMapper)}`
  return calls.reduce(reducer, '')
}

module.exports = {prettyCalls}
