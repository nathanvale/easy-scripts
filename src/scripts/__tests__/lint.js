import cases from 'jest-in-case'
import {unquoteSerializer, winPathSerializer} from '../../helpers/serializers'

const {prettyCalls} = require('../../helpers/pretty-calls')

jest.mock('../../utils')

expect.addSnapshotSerializer(unquoteSerializer)
expect.addSnapshotSerializer(winPathSerializer)

let crossSpawnSyncMock, originalArgv, printMock, originalExit

cases(
  'lint',
  async ({setup = () => () => {}}) => {
    // beforeEach
    jest.resetModules()
    crossSpawnSyncMock = require('cross-spawn').sync
    originalExit = process.exit
    printMock = require('../../utils').print
    process.exit = jest.fn()
    const teardown = await setup()
    try {
      // tests
      await require('../lint')
      // eslint-disable-next-line no-useless-catch
    } catch (error) {
      throw error
    } finally {
      process.exit = originalExit
      await teardown()
    }
  },
  {
    'calls eslint CLI with default args': {
      setup: withDefault(setupWithArgs()),
    },
    'should exit gracefully when an error is thrown': {
      setup: withDefault(withThrownError(setupWithArgs())),
    },
    'should print eslint fail message': {
      setup: withDefault(withEslintFail(setupWithArgs())),
    },
    'should compile files with --config and --ignore-path when using built in configs': {
      setup: withDefault(withBuiltInConfig(setupWithArgs())),
    },
    '--no-cache will disable caching': {
      setup: withDefault(setupWithArgs(['--no-cache'])),
    },
    'runs on given files, but only js,ts,tsx files': {
      setup: withDefault(
        setupWithArgs([
          './src/index.js',
          './src/index.ts',
          './src/index.tsx',
          './package.json',
          './src/index.css',
        ]),
      ),
    },
  },
)

function withBuiltInConfig(setupFn) {
  return function setup() {
    const {
      useBuiltInEslintConfig: useBuiltInEslintConfigMock,
      useBuiltInEslintIgnore: useBuiltInEslintIgnoreMock,
      fromConfigs,
    } = require('../../utils')
    useBuiltInEslintConfigMock.mockReturnValue(true)
    useBuiltInEslintIgnoreMock.mockReturnValue(true)
    fromConfigs.mockReturnValueOnce('~/src/config/eslintrc.js')
    fromConfigs.mockReturnValueOnce('~/src/config/eslintignore')
    const teardownFn = setupFn()
    return function teardown() {
      teardownFn()
    }
  }
}

function withThrownError(setupFn) {
  return function setup() {
    const error = new Error('some crazy error')
    crossSpawnSyncMock.mockImplementation(() => {
      throw error
    })
    const teardownFn = setupFn()

    return function teardown() {
      expect(prettyCalls(printMock.mock.calls)).toMatchInlineSnapshot(`
        Call 1:
          Argument 1:
            Linting FAILED ¯_(ツ)_/¯
        Call 2:
          Argument 1:
            Error: some crazy error
      `)
      teardownFn()
    }
  }
}

function withDefault(setupFn) {
  return function setup() {
    crossSpawnSyncMock.mockImplementation(() => ({
      status: 0,
    }))
    const teardownFn = setupFn()
    return function teardown() {
      expect(crossSpawnSyncMock).toHaveBeenCalledTimes(1)
      const [firstCall] = crossSpawnSyncMock.mock.calls
      const [script, calledArgs] = firstCall
      expect([script, ...calledArgs].join(' ')).toMatchSnapshot()
      teardownFn()
    }
  }
}

function withEslintFail(setupFn) {
  return function setup() {
    crossSpawnSyncMock.mockImplementation(() => ({
      status: 1,
    }))
    const teardownFn = setupFn()
    return function teardown() {
      expect(prettyCalls(printMock.mock.calls)).toMatchInlineSnapshot(`
        Call 1:
          Argument 1:
            Linting FAILED ¯_(ツ)_/¯
      `)
      teardownFn()
    }
  }
}

function setupWithArgs(args = []) {
  return function setup() {
    const utils = require('../../utils')
    utils.resolveBin = (modName, {executable = modName} = {}) => executable
    originalArgv = process.argv
    process.argv = ['node', '../lint', ...args]
    return function teardown() {
      process.argv = originalArgv
    }
  }
}
