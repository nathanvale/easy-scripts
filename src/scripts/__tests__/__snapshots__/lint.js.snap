// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lint --no-cache will disable caching 1`] = `eslint --no-cache --ext .js,.jsx,.ts,.tsx .`;

exports[`lint calls eslint CLI with default args 1`] = `eslint --cache --ext .js,.jsx,.ts,.tsx .`;

exports[`lint runs on given files, but only js files 1`] = `eslint --cache ./src/index.js ./src/component.js --ext .js,.jsx,.ts,.tsx`;

exports[`lint should compile files with --presets and --ignore-path when using built in configs 1`] = `eslint --config ~/src/config/eslintrc.js --ignore-path ~/src/config/eslintignore --cache --ext .js,.jsx,.ts,.tsx .`;
